### DISCUSSION TITLE
Support for user-defined foreign keys?

### DISCUSSION THREAD

**Discussion by jechterhoff** [ngageoint/geopackage-core-java#48] [General]
Hello,

Does the API support the creation of user-defined foreign keys? For example, a column in an attributes-table defined with foreign key constraint that references a particular features-table? 

The standards-compliant solution would be to use the GeoPackage Related-Tables-Extension, but some communities appear to be using foreign keys to encode relationships within a GeoPackage. So I am wondering if the NGA GeoPackage Java API would support that particular encoding approach.

Best regards,
Johannes Echterhoff

**Comment by bosborn** [comment]
That would be possible if including a raw constraint when creating the attributes table (from [geopackage-java](https://github.com/ngageoint/geopackage-java) or [geopackage-android](https://github.com/ngageoint/geopackage-android)).  Enforcement would then provided by SQLite.  There would be nothing restricting reading a GeoPackage set up that way.

```
Constraints constraints = new Constraints();
constraints.add(new RawConstraint(ConstraintType.FOREIGN_KEY, fk_name, 
		"FOREIGN KEY (attributes_column) REFERENCES feature_table (feature_column)"));

AttributesTable attributesTable = geoPackage.createAttributesTable(
		AttributesTableMetadata.create(attributes_table, columns, constraints));
```

**Reply by jechterhoff** [reply]
Thank you very much, Brian!
